cirq/contrib/acquaintance/strategy.py

```diff
@@ -0,0 +1,135 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+from functools import partial
+from operator import ne
+from typing import Sequence, TYPE_CHECKING, Union
+
+from cirq.circuits import Circuit, Moment, ExpandComposite
+from cirq.extension import Extensions
+from cirq.ops import QubitId, GateOperation
+from cirq.contrib.acquaintance.gates import (
+     SwapNetworkGate, AcquaintanceOpportunityGate, ACQUAINT,
+     op_acquaintance_size)
+from cirq.contrib.acquaintance.permutation import (
+        PermutationGate)
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    from typing import Type, Dict, List
+    from cirq.ops import Gate
+
+STRATEGY_GATE = Union[AcquaintanceOpportunityGate, PermutationGate]
+
+class AcquaintanceStrategy(Circuit):
+    """An acquaintance strategy is a special type of circuit that contains only
+    gates of two types: AcquaintanceOpportunityGate and PermutationGate. It
+    serves as a skeleton for compiling around geometric constraints.
+    """
+
+    gate_types = (AcquaintanceOpportunityGate, PermutationGate)
+
+    def __init__(self, circuit: Circuit=None) -> None:
+        if circuit is None:
+            circuit = Circuit()
+        for moment in circuit:
+            for op in moment.operations:
+                if not isinstance(op.gate, self.gate_types):
+                    raise ValueError('An acquaintance strategy can only'
+                        'contain gates of types {}'.format(self.gate_types))
+        super().__init__(circuit._moments)
+
+    def rectify(self, acquaint_first: bool=True):
+        """Splits moments so that they contain either only acquaintance gates
+        or only permutation gates."""
+        last_gate_type = self.gate_types[int(not acquaint_first)]
+        rectified_moments = []
+        for moment in self:
+            gate_type_to_ops = defaultdict(list
+                    ) # type: Dict[Type[STRATEGY_GATE], List[GateOperation]]
+            for op in moment.operations:
+                gate_type_to_ops[type(op.gate)].append(op)
+            if len(gate_type_to_ops) == 1:
+                rectified_moments.append(moment)
+                continue
+            for gate_type in sorted(gate_type_to_ops,
+                                    key=partial(ne, last_gate_type)):
+                rectified_moments.append(Moment(gate_type_to_ops[gate_type]))
+                last_gate_type = gate_type
+        self._moments = rectified_moments
+
+    def nest(self, qubit_order: Sequence[QubitId], acquaintance_size: int=0
```

Comment:
You should use the `QubitOrderOrList` type if possible, for consistency with other parts of the code.

Reply:
Here I'm using `qubit_order` not just for the ordering but also for the set of qubits; if I understand `QubitOrder` correctly, that would be insufficient here.